--- a/build_tools/menuselect-deps.in
+++ b/build_tools/menuselect-deps.in
@@ -44,2 +44,3 @@
 OPUS=@PBX_OPUS@
+OPUSENC=@PBX_OPUSENC@
 OPUSFILE=@PBX_OPUSFILE@
--- a/configure.ac
+++ b/configure.ac
@@ -523,2 +523,3 @@
 AST_EXT_LIB_SETUP([OPUS], [Opus], [opus])
+AST_EXT_LIB_SETUP([OPUSENC], [Opusenc], [opusenc])
 AST_EXT_LIB_SETUP([OPUSFILE], [Opusfile], [opusfile])
@@ -2524,2 +2525,3 @@
 fi
+AST_PKG_CONFIG_CHECK(OPUSENC, libopusenc)
 AST_EXT_LIB_CHECK([OPUSFILE], [opusfile], [op_open_callbacks], [opus/opusfile.h], [], [$__opus_include])
--- a/makeopts.in
+++ b/makeopts.in
@@ -228,6 +228,9 @@
 OPUS_INCLUDE=@OPUS_INCLUDE@
 OPUS_LIB=@OPUS_LIB@
 
+OPUSENC_INCLUDE=@OPUSENC_INCLUDE@
+OPUSENC_LIB=@OPUSENC_LIB@
+
 OPUSFILE_INCLUDE=@OPUSFILE_INCLUDE@
 OPUSFILE_LIB=@OPUSFILE_LIB@
 
--- formats/Makefile
+++ formats/Makefile
@@ -12,4 +12,6 @@
 -include $(ASTTOPDIR)/menuselect.makeopts $(ASTTOPDIR)/menuselect.makedeps
 
+_ASTCFLAGS+=-DASTERISK_VERSION_NUM=${ASTERISKVERSIONNUM}
+
 MODULE_PREFIX=format
 MENUSELECT_CATEGORY=FORMATS
